File: data_structures_and_algorithms/Code_401/class-11/solutions/python/pseudo-queue/pseudo_queue/pseudo_queue.py
class PseudoQueue:
    def __init__(self):
        self._inbox = Stack()
        self._outbox = Stack()

    def enqueue(self, value):

        while not self._outbox.is_empty():
            self._inbox.push(self._outbox.pop())

        self._inbox.push(value)

    def dequeue(self):
        
        while not self._inbox.is_empty():
            self._outbox.push(self._inbox.pop())

        if self._outbox.is_empty():
            raise RuntimeError("Cannot dequeue from empty queue")

        return self._outbox.pop()
    

class Stack:
    def __init__(self):
        self.top = None

    def push(self, value):
        self.top = Node(value, self.top)

    def pop(self):

        if not self.top:
            raise RuntimeError("Cannot pop from empty stack")

        popped = self.top
        self.top = self.top.next
        return popped.value

    def peek(self):

        if not self.top:
            raise RuntimeError("Cannot peek on empty stack")

        return self.top.value

    def is_empty(self):
        return self.top is None


class Node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next

File: data_structures_and_algorithms/Code_401/class-11/solutions/java/PseudoQueue.java
import java.util.*;

class PseudoQueue <T> {
    public static void main(String[] args) throws Exception{
        PseudoQueue<Integer> queue1 = new PseudoQueue<Integer>();
        queue1.enqueue(1);
        queue1.enqueue(3);
        queue1.enqueue(5);
        System.out.println("queue1.isEmpty() = " + queue1.isEmpty());
        System.out.println("queue1.dequeue() = " + queue1.dequeue());
        System.out.println("queue1.peek() = " + queue1.peek());
        System.out.println("queue1.dequeue() = " + queue1.dequeue());
        System.out.println("queue1.dequeue() = " + queue1.dequeue());
        System.out.println("queue1.isEmpty() = " + queue1.isEmpty());

    }

    Stack<T> inStack = new Stack();
    Stack<T> outStack = new Stack();



    public void enqueue(T value){
        inStack.push(value);
    }

    public T dequeue() throws Exception{
        if(outStack.isEmpty()){
            while(!inStack.isEmpty()){
                outStack.push(inStack.pop());
            }
            if(outStack.isEmpty()) throw new Exception("PsuedoQueue is empty");
        }

        return outStack.pop();
    }

    public T peek() throws Exception{
        if(outStack.isEmpty()){
            while(!inStack.isEmpty()){
                outStack.push(inStack.pop());
            }
            if(outStack.isEmpty()) throw new Exception("PsuedoQueue is empty");
        }

        return outStack.peek();
    }

    public boolean isEmpty(){
        if(outStack.isEmpty() && inStack.isEmpty()) return true;
        return false;
    }

}

